File: update_universe.js
const fs = require('fs');
const https = require('https');

const url = 'https://nsearchives.nseindia.com/content/equities/EQUITY_L.csv';
const universePath = 'web/public/universe.json';

https.get(url, { headers: { 'User-Agent': 'Mozilla/5.0' } }, (res) => {
  let data = '';
  res.on('data', chunk => data += chunk);
  res.on('end', () => {
    // Parse existing universe to keep sectors
    let existingUniverse = [];
    try {
      existingUniverse = JSON.parse(fs.readFileSync(universePath, 'utf8'));
    } catch (e) {
      console.log('No existing universe.json found.');
    }
    
    const existingSymbols = new Map();
    existingUniverse.forEach(item => existingSymbols.set(item.symbol, item));

    const lines = data.split('\n');
    const newUniverse = [];
    
    // Skip header
    for (let i = 1; i < lines.length; i++) {
      const line = lines[i];
      if (!line.trim()) continue;
      
      // SYMBOL,NAME OF COMPANY, SERIES, DATE OF LISTING, PAID UP VALUE, MARKET LOT, ISIN NUMBER, FACE VALUE
      // e.g. 20MICRONS,20 Microns Limited,EQ,06-OCT-2008,5,1,INE144J01027,5
      // Note: Commas can be inside quotes for company names.
      // A simple split by comma might break if company name has comma, but NSE EQUITY_L generally avoids commas in company names, or quotes them.
      // Let's use a basic CSV regex split.
      const cols = line.split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/);
      if (cols.length < 3) continue;
      
      const symbol = cols[0].trim();
      const name = cols[1].replace(/^"|"$/g, '').trim();
      const series = cols[2].trim();
      
      if (series === 'EQ' || series === 'BE' || series === 'SM') {
        if (existingSymbols.has(symbol)) {
          newUniverse.push(existingSymbols.get(symbol));
        } else {
          newUniverse.push({
            symbol: symbol,
            name: name,
            sector: "Equities" // Default sector for new stocks
          });
        }
      }
    }
    
    // Add any existing that might have been skipped (indices, ETFs)
    const newSymbolsSet = new Set(newUniverse.map(s => s.symbol));
    existingUniverse.forEach(item => {
      if (!newSymbolsSet.has(item.symbol)) {
        newUniverse.push(item);
      }
    });
    
    // Sort alphabetically
    newUniverse.sort((a, b) => a.symbol.localeCompare(b.symbol));
    
    fs.writeFileSync(universePath, JSON.stringify(newUniverse, null, 2));
    console.log(`Successfully generated universe.json with ${newUniverse.length} stocks!`);
  });
}).on('error', err => {
  console.error('Error fetching NSE list:', err.message);
});

File: update_universe.py
import pandas as pd
import requests
import io
import os
import ast
import re

url = "https://nsearchives.nseindia.com/content/equities/EQUITY_L.csv"
headers = {"User-Agent": "Mozilla/5.0"}
res = requests.get(url, headers=headers)
df = pd.read_csv(io.StringIO(res.text))

df = df[df[' SERIES'].str.strip() == 'EQ']

with open('alphapulse/universe.py', 'r', encoding='utf-8') as f:
    old_content = f.read()

match_universe = re.search(r'NSE_UNIVERSE:\s*list\[dict\]\s*=\s*(\[.*?\])', old_content, re.DOTALL)
if match_universe:
    existing_universe = ast.literal_eval(match_universe.group(1))
else:
    existing_universe = []

match_nifty = re.search(r'NIFTY_50_SYMBOLS:\s*list\[str\]\s*=\s*(\[.*?\])', old_content, re.DOTALL)
nifty_50_text = match_nifty.group(1) if match_nifty else "[]"

existing_symbols = {item['symbol']: item for item in existing_universe}

new_universe = []
for index, row in df.iterrows():
    sym = row['SYMBOL'].strip()
    name = row['NAME OF COMPANY'].strip()
    
    if sym in existing_symbols:
        new_universe.append(existing_symbols[sym])
    else:
        new_universe.append({
            "symbol": sym,
            "name": name,
            "sector": "General"
        })

new_symbols_set = set([s['symbol'] for s in new_universe])
for sym, data in existing_symbols.items():
    if sym not in new_symbols_set:
        new_universe.append(data)

new_universe = sorted(new_universe, key=lambda x: x['symbol'])

print(f"Successfully fetched and merged. Total stocks: {len(new_universe)}")

new_py = '"""Comprehensive NSE 2000+ Universe covering all listed equities on the National Stock Exchange."""\n\n'
new_py += 'NSE_UNIVERSE: list[dict] = [\n'
for item in new_universe:
    name_escaped = item['name'].replace("'", "\\'")
    new_py += f'    {{"symbol": "{item["symbol"]}", "name": "{name_escaped}", "sector": "{item["sector"]}"}},\n'
new_py += ']\n\n'

new_py += f'NIFTY_50_SYMBOLS: list[str] = {nifty_50_text}\n\n'
new_py += 'def get_all_tickers() -> list[str]:\n'
new_py += '    return [s["symbol"] + ".NS" for s in NSE_UNIVERSE]\n\n'
new_py += 'def get_by_sector(sector: str) -> list[dict]:\n'
new_py += '    return [s for s in NSE_UNIVERSE if s["sector"] == sector]\n'

with open('alphapulse/universe.py', 'w', encoding='utf-8') as f:
    f.write(new_py)
